The-SMIRNOFF-force-field-format.md

```diff
@@ -1,103 +1,288 @@
 ```XML
-   <HarmonicAngleForce angle_unit="degrees" k_unit="kilocalories_per_mole/radian**2">
+<AngleForce angle_unit="degrees" k_unit="kilocalories_per_mole/radian**2">
+   ...
+</AngleForce>     
 ```
 which introduces following `Angle` terms which will use units of degrees for the angle and kilocalories per mole per square radian for the force constant.
 
+**QUESTION:** Can we change `smirks` to `smarts` in the attributes, since these are really SMARTS strings augmented with atom tagging (which seems to be broadly supported)?
+
 Under each of these force terms, there are tags for individual parameter lines such as these:
 ```XML
+<AngleForce angle_unit="degrees" k_unit="kilocalories_per_mole/radian**2">
    <Angle smirks="[a,A:1]-[#6X4:2]-[a,A:3]" angle="109.50" k="100.0"/>
    <Angle smirks="[#1:1]-[#6X4:2]-[#1:3]" angle="109.50" k="70.0"/>
+</AngleForce>     
 ```
 The first of these specifies the `[a,A:1]-[#6X4:2]-[a,A:3]` SMIRKS pattern for an angle, with a tetravalent carbon at the center with single bonds to two atoms of any type.
-Atoms are labeled 1, 2, and 3, with 2 being the central atom. Equilibrium angle values are provided, along with force constants (with units as given above).
+Atoms are labeled 1, 2, and 3, with 2 being the central atom.
+Equilibrium angle values are provided, along with force constants (with units as given above).
+
+Before getting in to individual sections, it's worth noting that the XML parser ignores attributes in the XML that it does not understand, so providing a parameter line for an angle that specifies (for example) a second force constant `k2` will lead to no effect.
 
 **SMIRNOFF parameters are hierarchical** in that parameters which come later in a file override those which come earlier if they match the same pattern.
  This can be seen in the example above, where the first line provides a generic angle parameter for any tetravalent carbon (single bond) angle, and the second line overrides this for the specific case of a hydrogen-(tetravalent carbon)-hydrogen angle.
 
 This hierarchical structure means that a typical parameter file will tend to have generic parameters early in the section for each force type, with more specialized parameters assigned later.
 
+Multiple `.ffxml` files can be loaded in sequence.
+If these files each contain unique `<...Force>` tags, the resulting forcefield will be independent of the order in which the files are loaded.
+If, however, the same tag occurs in multiple files, the contents of the tags are merged, with the tags read later appended to the *end* of the parameters read earlier, provided the `<...Force>` tags have compatible attributes.
+The resulting forcefield will therefore depend on the order in which parameters are read.
```

Comment:
The merging doesn't quite make sense to me, but let me see if this is a correct interpretation: If I load a general `ffxml` first, that contains an angle parameter for `O-H-O`, and then I load a specific water `ffxml`, also with parameters for with `O-H-O`, then the tags (and parameters) will be merged, unless I make the water `ffxml` more specific so the water molecules in my system don't match the generic `O-H-O` angle in the first `ffxml`? 

I have a feeling one "risk" of using a SMIRNOFF format force field (as a user) is thinking that we know which set of parameters are being used and not realizing something else is actually being matched (or merged). Is there a goal to add some sort of logging to `createSystem` that will print out which parameters are being matched (I know David / Caitlin have a notebook that can highlight parts of a molecule that match certain SMARTS/SMIRKS patterns, but I don't imagine most people setting up systems will go through that in a typical workflow)?

Reply:
Yes, I think this specific phrase needs a little clarification:
> If, however, the same tag occurs in multiple files, the contents of the tags are merged, with the tags read later appended to the *end* of the parameters read earlier, provided the `<...Force>` tags have compatible attributes.

It would be easy to read this as saying, if you provide several sets of parameters for the same SMIRKS pattern, you will end up with all of those parameters being applied (e.g., if you provide two files with a `[#6X4]-[#6X4]` bond parameter, you'll end up with a single line which applies both force constants and both bond distances). That doesn't make sense, but to me it's the most obvious reading. 

Maybe we can make this clear just by being more concrete: "If, however, the same force tag occurs in multiple files, the contents of these tags are merged, with.... For example, if a `BondForce` tag is read from two files, the resulting `BondForce`s in the result will first include bond parameters from the first file then those from the second file. This is in keeping with the hierarchical format generally employed here, and allows (for example) a specialized water model to be read last and override any parameters which would otherwise be assigned to water. However, it is important to note that when multiple files are employed, the resulting forcefield will therefore depend on the order in which parameters are read."

@slochower - does that help? And yes, I think there's a risk. We do have a `labelSystem` option that basically does exactly what you're saying (logs what parameters are applied where` and it exactly parallels `createSystem`.

One long-term goal I think is for us to build complete consistent force fields which are contained in single files. But in the interim I think we're stuck with the current world where users expect to be able to pick, say, a small molecule force field, a water model, and an ion parameter set and mix and match, which creates a need to do this kind of thing.

Reply:
> It would be easy to read this as saying, if you provide several sets of parameters for the same SMIRKS pattern, you will end up with all of those parameters being applied (e.g., if you provide two files with a [#6X4]-[#6X4] bond parameter, you'll end up with a single line which applies both force constants and both bond distances). That doesn't make sense, but to me it's the most obvious reading.

That's exactly the ambiguity I was thinking about. 

> Maybe we can make this clear just by being more concrete: "If, however, the same force tag occurs in multiple files, the contents of these tags are merged, with.... For example, if a BondForce tag is read from two files, the resulting BondForces in the result will first include bond parameters from the first file then those from the second file. This is in keeping with the hierarchical format generally employed here, and allows (for example) a specialized water model to be read last and override any parameters which would otherwise be assigned to water. However, it is important to note that when multiple files are employed, the resulting forcefield will therefore depend on the order in which parameters are read."

I think this is good if it is explicit what "merging" means because I'm still not clear what happens if the same parameter with the same force constants is included (perhaps by mistake) in two files versus the same parameter with different force constants in two files.

I also think it would be nice to put `labelSystem` somewhere up front in the documentation (but perhaps not here in the specifications); it could really help with on boarding. Also, perhaps, including or linking to Figure 1 (I think?) from the current SMIRNOFF paper, which just shows pattern matching. I'm pretty sure you also have some suitable slides from at least one of your (@davidlmobley ) presentations.

Reply:
If the same parameter is included by mistake with the same or different force constants, the last one would win, so the first one would do nothing.

Reply:
Okay, that's fine -- but that doesn't sound like what I would call "merging" or "appending". It sounds like replacement. Maybe I'm still misunderstanding something about what is happening here.

Reply:
It's "appending"/"merging" in the sense that both parameters will be present in the XML and in the forcefield, but since they are equivalent and the format is hierarchical, the "last one wins" approach would mean only the last one will ever be applied. 

I suppose we could also just DELETE identical SMIRKS patterns, but that only removes a subset of the problem -- e.g. `[#1]` and `[#1H0]` and `[#1X1+0]` etc. are all going to match the same chemistry (except for the H_2 molecule) but they're not identical, so they'd multiple of these would end up present if you combined force fields which matched hydrogen in different ways. And only the last one would be applied.

Wording suggestions to clarify this are welcomed.

Reply:
Maybe something like this?

"Multiple .ffxml files can be loaded in sequence. If these files each contain unique <...Force> tags, the resulting forcefield will be independent of the order in which the files are loaded. 

If, however, the same tag occurs in multiple files, the contents of the tags are merged, with the tags read later taking precedence over the parameters read earlier, provided the <...Force> tags have compatible attributes. The resulting force field will therefore depend on the order in which parameters are read. This behavior is intended for limited use in appending very specific parameters (such as solvent models)."

Reply:
LGTM.